src/main/cpp/LidarReader.cpp

```diff
@@ -0,0 +1,114 @@
+#include "LidarReader.h"
+
+#include <iostream>
+#include <algorithm>
+
+#include "frc/smartdashboard/SmartDashboard.h"
+
+LidarReader::LidarReader():
+    port_(LidarReaderConstants::BAUD_RATE, LidarReaderConstants::LIDAR_PORT)
+{
+    frc::SmartDashboard::PutBoolean("Lidar Responding", true);
+}
+
+/// @brief Requests Lidar for data
+void LidarReader::RequestData(){
+    if(isRequesting_){
+        return;
+    }
+    port_.Write(LidarReaderConstants::REQ, 1);
+    reqTime_ = frc::Timer::GetFPGATimestamp().value();
+    isRequesting_ = true;
+}
+
+/// @brief Should be called in perioidic - reads port and stores data
+void LidarReader::Periodic(){
+    //If not requesting, do nothing
+    if(!isRequesting_){
+        return;
+    }
+    //Futher logic assumes a request is in place
+
+    //Check if it is responding in time
+    if(frc::Timer::GetFPGATimestamp().value() - reqTime_ > LidarReaderConstants::RESPONSE_TIME){
+        data_.isValid = false;
+        isRequesting_ = false;
+        port_.Reset();
+        RequestData();
+    }
+    else{
+        data_.isValid = true;
+    }
+    frc::SmartDashboard::PutBoolean("Lidar Responding", data_.isValid);
+
+    //Check buffer size
+    int bufferSize = port_.GetBytesReceived();
+    if(bufferSize == 0){
```

Comment:
validity is checked only after reading 4 bytes. I wouldn't check validity here because all 4 bytes might not have been sent while this periodic call runs.